File: packages/shared/src/api/agents.ts
import { z } from "zod";

import { agentProbeSchema } from "../agent-capabilities.js";
import { AGENT_CAPABILITIES, AGENT_IDS, PROMPT_MODES } from "../agents.js";
import { agentIdSchema } from "../domain/index.js";

/** A `LaunchWord` (`shell.ts`): a literal value, or a prefix and a variable. */
export const launchWordSchema = z.union([
  z.string(),
  z.object({ prefix: z.string(), env: z.string() }),
]);

/** `GET /api/agents` — registry entry plus local detection. */
export const agentInfoSchema = z.object({
  id: agentIdSchema,
  label: z.string(),
  color: z.string(),
  launchCmd: z.string(),
  promptMode: z.enum(PROMPT_MODES),
  capabilities: z.array(z.enum(AGENT_CAPABILITIES)).default([]),
  /** Extra argv the launch line appends; always empty for a built-in agent. */
  args: z.array(z.string()).default([]),
  /**
   * The built-in agent a `custom:` entry borrows its hooks, prompt mode and
   * permission flags from (plan §24.1). Absent on the built-ins themselves.
   */
  baseAgent: z.enum(AGENT_IDS).optional(),
  /** Absolute path the launch program resolves to, or null. */
  resolvedPath: z.string().nullable().default(null),
  /** The launch program exists on the augmented PATH. */
  installed: z.boolean(),
  /**
   * What to start instead of `resolvedPath` when that is an npm / pnpm
   * wrapper on Windows (`claude.cmd`): the program behind it and the words
   * that go in front of the CLI's own (`node.exe <cli.js>`). A batch wrapper
   * has `cmd.exe` read every argument again, which no quoting survives.
   * Absent when the path is the program itself or the wrapper was not read.
   */
  launchTarget: z
    .object({ program: z.string(), args: z.array(z.string()) })
    .optional(),
  /** Revision of the installed hook client, absent when hooks are not installed. */
  clientRevision: z.number().int().nonnegative().nullish(),
  /**
   * Revision of the installed collaboration skill, absent when the skill is not
   * installed. It is read from the file on disk rather than from a row, so a
   * user who deletes the skill by hand sees that here on the next refresh.
   */
  skillsRevision: z.number().int().nonnegative().nullish(),
  /**
   * Cached `--version` probe (`agent-capabilities.ts`). Absent means the CLI
   * has not been probed yet, which resolves gated capabilities to `unknown` —
   * never to supported.
   */
  probe: agentProbeSchema.nullish(),
  /**
   * Argv a canvas launch of this agent carries — the canvas injection
   * (docs/design/canvas-only-integration.md §2): Claude's `--settings` /
   * `--plugin-dir` / `--append-system-prompt-file`, Codex's `-c` hooks and
   * developer instructions, and so on. Empty when nothing is prepared.
   *
   * It is answered per request rather than frozen into a launch definition
   * because both halves are the runtime's: the path is that data directory's
   * and the flag is that CLI version's.
   *
   * Optional rather than defaulted: the runtime omits it when there is nothing
   * to add, and "this agent needs no argv" and "this runtime predates the
   * field" are the same instruction to the caller — add nothing.
   */
  launchArgs: z.array(z.string()).optional(),
  /**
   * The same injection as words for a typed launch line, not yet quoted: the
   * page quotes them for the node terminal's shell (`shell.ts`). Codex's
   * words name environment variables the node's terminal carries
   * (`{ prefix, env }`) instead of spelling kilobytes out: a line that long is
   * cut off while a fresh shell is still echoing it. The page uses these when
   * present, and `launchArgs` otherwise.
   */
  launchWords: z.array(launchWordSchema).optional(),
});

export const agentListSchema = z.array(agentInfoSchema);

/**
 * `GET /api/agents/{id}/models` — 节点头部「模型」菜单的候选（用户实测反馈 F7）。
 *
 * `source` 说的是这一条**从哪来**，因为三种来源不是一回事：
 *
 *   * `cli` —— CLI 自己说的（`claude --help` 的别名、Codex `config.toml` 里
 *     配好的模型）。它反映的是这台机器、这个账号的实际情况，所以排在最前；
 *   * `catalog` —— models.dev 上该 provider 的条目，按发布日期倒序；
 *   * `builtin` —— 离线兜底，只在前两者都拿不到时出现。
 *
 * 没有 `releaseDate` 的条目排在有日期的之前：别名永远指向该系列的最新模型，
 * 不可能比下面任何一条更旧，而且那是 CLI 自己文档里的写法。
 */
export const agentModelSourceSchema = z.enum(["cli", "catalog", "builtin"]);
export type AgentModelSource = z.infer<typeof agentModelSourceSchema>;

export const agentModelSchema = z.object({
  /** 原样放到启动行 `--model` 后面的值。 */
  id: z.string().min(1),
  label: z.string(),
  source: agentModelSourceSchema,
  /** `YYYY-MM-DD`，目录里有才有。 */
  releaseDate: z.string().optional(),
});
export type AgentModel = z.infer<typeof agentModelSchema>;

export const agentModelListSchema = z.array(agentModelSchema);

/**
 * How a provider's integration reaches its CLI.
 *
 * `canvas` is the only mode a current core answers
 * (docs/design/canvas-only-integration.md): hook, skill and canvas
 * instructions are handed over on the launch line of a canvas node and nowhere
 * else. The older three stay parseable so a page talking to an older core
 * still draws its rows.
 */
export const INJECTION_MODES = [
  "canvas",
  "launch",
  "file",
  "extension",
] as const;

/** One half of an install unit — the adapter, or the skill. */
export const integrationHalfSchema = z.object({
  installed: z.boolean(),
  /** Where it lives. Present even when it is not installed: the answer to
   * "why is this not on" is usually "look here". */
  path: z.string().optional(),
  /** The revision on disk; `0` when this half is not installed. */
  revision: z.number().int().nonnegative().default(0),
});

/** Something an earlier product name left behind (设计 §4). */
export const legacyIntegrationFindingSchema = z.object({
  /** `hook_entry` / `skill_dir` / `codex_unknown_key` / `status_line` / `instruction_block`. */
  kind: z.string(),
  path: z.string(),
  /** The command, key or directory name, so a person can recognise their own. */
  detail: z.string(),
});

/**
 * `GET /api/agents/{id}/integration`, and what install / uninstall answer with
 * (docs/design/agent-integration.md §5).
 *
 * Hook and skill are **one** install unit with one state: before this, a CLI
 * had two switches and three ways to be half-integrated, and no single screen
 * could say which. `revision` is what a fresh install writes — the hook
 * revision and the skill revision folded together — and `stale` is the only
 * question the page has to ask about it.
 *
 * Loose on purpose: the runtime omits what is empty and adds fields faster than
 * the settings page reads them.
 */
export const integrationStateSchema = z.looseObject({
  agentId: agentIdSchema,
  mode: z.enum(INJECTION_MODES),
  hook: integrationHalfSchema,
  skill: integrationHalfSchema,
  legacy: z.object({
    found: z.array(legacyIntegrationFindingSchema).default([]),
  }),
  revision: z.number().int().nonnegative(),
  /** What the files on disk were written by; absent when nothing is installed. */
  installedRevision: z.number().int().nonnegative().optional(),
  stale: z.boolean().default(false),
  /**
   * Argv a session of this agent must carry for its adapter to load. Empty for
   * every mode but `launch`, and for an integration that is not installed.
   */
  launchArgs: z.array(z.string()).default([]),
  /** The same as words for a typed launch line, unquoted. */
  launchWords: z.array(launchWordSchema).default([]),
  /** Names of the environment variables a canvas launch sets. */
  launchEnv: z.array(z.string()).default([]),
  /**
   * Files outside the data directory this integration writes — Codex's
   * `config.toml`, for its hook trust records, and nothing else.
   */
  globalWrites: z.array(z.string()).default([]),
  /** What the one-time move away from the old global install did here. */
  migration: z
    .object({
      migratedAt: z.string(),
      removed: z.array(z.string()).default([]),
      backups: z.array(z.string()).default([]),
      error: z.string().optional(),
    })
    .optional(),
  clientBin: z.string().optional(),
  /** Something worked but deserves a sentence in the settings page. */
  warning: z.string().optional(),
});

/**
 * `POST /api/agents/{id}/integration/repair` — what the pass actually did.
 *
 * `kept` is the point of the report: everything this repair recognised as not
 * ours and wrote back exactly as it read it.
 */
export const integrationRepairReportSchema = z.looseObject({
  agentId: agentIdSchema,
  found: z.array(legacyIntegrationFindingSchema).default([]),
  removed: z.array(z.string()).default([]),
  kept: z.array(z.string()).default([]),
  /** The newest backup, for the sentence the settings page shows. */
  backup: z.string().optional(),
  backups: z.array(z.string()).default([]),
});

export const answerApprovalRequestSchema = z.object({
  decision: z.enum(["allow", "deny"]),
});

export const answerApprovalResponseSchema = z.object({
  pendingId: z.string(),
  decision: z.enum(["allow", "deny"]),
  answeredAt: z.string().datetime({ offset: true }),
});

/**
 * `POST /api/agent-status/{nodeId}/suggest-title` — the header's ✦ button.
 *
 * `source` says where the sentence came from so the UI can be honest when the
 * answer is only the agent's name: `transcript` (first user message),
 * `terminal` (last command in the pane) or `agent` (the label, nothing better
 * was available).
 */
export const suggestTitleResponseSchema = z.object({
  title: z.string().min(1).max(40),
  source: z.enum(["transcript", "terminal", "agent"]),
});

/**
 * `GET /api/agent-status/{nodeId}/transcript` — the node's own conversation,
 * one prose line per message.
 *
 * `truncated` is a field rather than an ellipsis in the text because a cut-off
 * conversation read as a whole one is a wrong answer, not a short one. A
 * provider that keeps nothing readable answers 501 with a reason, so an empty
 * `text` here never stands for "this CLI has no transcript".
 */
export const agentTranscriptSchema = z.object({
  nodeId: z.string().max(160),
  text: z.string(),
  truncated: z.boolean(),
});

/**
 * What a linked whiteboard shape reads as (docs/design/canvas-react-flow.md
 * §2.5). Only present when `kind === "shape"`: text items carry their text,
 * everything else is rasterised by the client and referenced by a
 * workspace-relative PNG path.
 */
export const contextLinkContentSchema = z.object({
  /** Render status is explicit: a visible link need not have a ready image. */
  status: z.enum(["pending", "ready", "error"]).optional(),
  sourceShapeId: z.string().max(160).optional(),
  shapeType: z.string().max(40).optional(),
  textTruncated: z.boolean().optional(),
  text: z.string().max(20_000).optional(),
  pngPath: z.string().max(4_000).optional(),
});

export const contextLinkSchema = z.object({
  /** Node id, or the uuid of the whiteboard item behind a `shape` link. */
  id: z.string().uuid(),
  title: z.string().max(160),
  /**
   * A node type, or `"shape"` for whiteboard content
   * (docs/design/canvas-react-flow.md §2.5).
   */
  kind: z.string().max(40),
  content: contextLinkContentSchema.optional(),
});

export type SuggestTitleResponse = z.infer<typeof suggestTitleResponseSchema>;
export type AgentTranscript = z.infer<typeof agentTranscriptSchema>;
export type AgentInfo = z.infer<typeof agentInfoSchema>;
export type InjectionMode = (typeof INJECTION_MODES)[number];
export type IntegrationHalf = z.infer<typeof integrationHalfSchema>;
export type LegacyIntegrationFinding = z.infer<
  typeof legacyIntegrationFindingSchema
>;
export type IntegrationState = z.infer<typeof integrationStateSchema>;
export type IntegrationRepairReport = z.infer<
  typeof integrationRepairReportSchema
>;
export type AnswerApprovalRequest = z.infer<typeof answerApprovalRequestSchema>;
export type ContextLink = z.infer<typeof contextLinkSchema>;
export type ContextLinkContent = z.infer<typeof contextLinkContentSchema>;

File: apps/runtime/src/api/agents.rs
//! `/api/agents`, `/api/conversations` and `/api/agent-status` — agent
//! discovery, the integration switch and the session status surfaces.

use std::path::Path;

use axum::{
    Json,
    extract::{Path as AxumPath, Query, State},
};
use serde::{Deserialize, Serialize};

use crate::{
    AppState,
    agent::{self, AgentInfo},
    agent_probe, db,
    error::{AppError, AppResult},
    events::WorkspaceEvent,
    hook::install::{integration, repair},
    index,
    model::{AgentStatus, Conversation},
    ownership,
};

/* ------------------------------- conversations ---------------------------- */

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConversationQuery {
    q: Option<String>,
    limit: Option<i64>,
}

/// `GET /api/conversations?q=&limit=50` — the command palette's history group
/// (plan §17). Newest first; `q` matches the title or the working directory,
/// case-insensitively, as a substring.
pub async fn list_conversations(
    State(state): State<AppState>,
    Query(query): Query<ConversationQuery>,
) -> AppResult<Json<Vec<Conversation>>> {
    Ok(Json(
        index::list(
            &state.pool,
            query.q.as_deref(),
            query.limit.unwrap_or(index::DEFAULT_LIMIT),
        )
        .await?,
    ))
}

/// `POST /api/conversations/refresh` — rescan now instead of waiting for the
/// 60 s timer. Same pass the timer runs, so calling it twice is harmless.
pub async fn refresh_conversations(
    State(state): State<AppState>,
) -> AppResult<Json<index::ScanReport>> {
    Ok(Json(index::refresh(&state.pool).await?))
}

/* ---------------------------------- agents -------------------------------- */

/// Registry mirror + local detection + hook install state.
///
/// The built-ins come first, then `settings.agents.custom[]` (plan §24.1). A
/// custom entry reports the hook install of the agent it borrows, because that
/// is the hook that will actually fire for it.
pub async fn agents(State(state): State<AppState>) -> AppResult<Json<Vec<AgentInfo>>> {
    let installs = db::list_hook_installs(&state.pool).await?;
    // The probe cache lives in the settings document, so it may only be written
    // while this Runtime still owns that domain; the probe itself is a fact
    // about this machine and runs either way.
    let persist =
        ownership::local_write_allowed(&state.pool, ownership::OwnershipDomain::Settings).await;
    let mut detected = agent::detect();
    detected.extend(
        state
            .settings
            .custom_agents()
            .iter()
            .map(agent::custom_info),
    );
    for info in &mut detected {
        let hook_provider = info.base_agent.unwrap_or(info.id.as_str());
        info.client_revision = installs
            .iter()
            .find(|install| install.agent_id == hook_provider)
            .map(|install| install.client_revision);
        // The skill install has no row of its own: the file on disk *is* the
        // state, so a user who deletes it by hand sees that here without
        // anything having to notice.
        info.skills_revision = crate::collab::skills::installed_revision_for(hook_provider);
        // What a session of this agent has to be launched with for its adapter
        // to load at all (设计 §3). Empty for every provider but Claude Code,
        // and for an integration that is not installed.
        info.launch_args = integration::launch_args(hook_provider);
        // Version probing is what decides whether a gated capability is
        // `supported` or `unknown` on the client (design §1). A program that
        // is not installed is not run: there is nothing to ask.
        if info.installed {
            info.probe = Some(
                agent_probe::cached(&state.settings, &info.id, &info.launch_cmd, persist).await,
            );
        }
    }
    Ok(Json(detected))
}

/// `GET /api/agents/{agent_id}/models` — what this CLI's model menu may offer
/// (用户实测反馈 F7).
///
/// Newest first, each entry saying whether the CLI itself named it, the
/// models.dev catalog did, or it is the offline fallback. A `custom:` entry
/// borrows its base adapter's list, the same way it borrows its hooks — but it
/// is probed through *its own* launch program, because that is the binary its
/// terminals will actually run.
pub async fn agent_models(
    State(state): State<AppState>,
    AxumPath(agent_id): AxumPath<String>,
) -> AppResult<Json<Vec<crate::models::agents::AgentModel>>> {
    let custom = state.settings.custom_agent(&agent_id);
    let (base, launch_cmd) = match &custom {
        Some(entry) => (entry.base_agent.clone(), entry.launch_cmd.clone()),
        None => {
            let definition = agent::definition(&agent_id)
                .ok_or_else(|| AppError::NotFound(format!("Unknown agent {agent_id}")))?;
            (definition.id.to_owned(), definition.launch_cmd.to_owned())
        }
    };
    Ok(Json(
        crate::models::agents::models_for(&base, Some(&launch_cmd)).await,
    ))
}

/* ------------------------------- integration ------------------------------ */

/// The provider whose files an integration request writes. A `custom:` entry
/// borrows its base agent's hooks *and* its skill directory: that is the
/// adapter that will actually fire for it (plan §24.1).
fn integration_provider(state: &AppState, agent_id: &str) -> String {
    state
        .settings
        .custom_agents()
        .iter()
        .find(|custom| custom.id == agent_id)
        .map(|custom| custom.base_agent.clone())
        .unwrap_or_else(|| agent_id.to_owned())
}

/// `GET /api/agents/{id}/integration` — one answer for one switch
/// (docs/design/agent-integration.md §5).
///
/// A read, so it answers under either ownership: the files are this machine's
/// and the Worker channel's `GetIntegration` reads the very same ones.
pub async fn read_integration(
    State(state): State<AppState>,
    AxumPath(agent_id): AxumPath<String>,
) -> AppResult<Json<integration::IntegrationState>> {
    Ok(Json(integration::state(&integration_provider(
        &state, &agent_id,
    ))?))
}

/// `POST /api/agents/{id}/integration/install` — hook and skill together.
/// Idempotent by construction; see `hook::install`.
pub async fn install_integration(
    State(state): State<AppState>,
    AxumPath(agent_id): AxumPath<String>,
) -> AppResult<Json<integration::IntegrationState>> {
    let provider = integration_provider(&state, &agent_id);
    let installed = integration::install(&provider)?;
    // The row is the Host's view of this machine's install state; the files on
    // disk remain the truth this Runtime reads back.
    db::upsert_hook_install(
        &state.pool,
        &provider,
        installed.hook.revision,
        installed.hook.path.as_deref(),
    )
    .await?;
    if let Some(warning) = &installed.warning {
        tracing::warn!(%agent_id, %warning, "the integration was installed with a caveat");
    }
    Ok(Json(installed))
}

pub async fn uninstall_integration(
    State(state): State<AppState>,
    AxumPath(agent_id): AxumPath<String>,
) -> AppResult<Json<integration::IntegrationState>> {
    let provider = integration_provider(&state, &agent_id);
    let removed = integration::uninstall(&provider)?;
    db::remove_hook_install(&state.pool, &provider).await?;
    Ok(Json(removed))
}

/// `POST /api/agents/{id}/integration/repair` — the only thing that edits what
/// an earlier product name left behind (设计 §4). Startup detects and reports;
/// this writes.
pub async fn repair_integration(
    State(state): State<AppState>,
    AxumPath(agent_id): AxumPath<String>,
) -> AppResult<Json<repair::RepairReport>> {
    let provider = integration_provider(&state, &agent_id);
    let report = repair::repair(&provider)?;
    if !report.removed.is_empty() {
        tracing::info!(
            agent = %provider,
            removed = report.removed.len(),
            backup = ?report.backup,
            "repaired an older install"
        );
    }
    Ok(Json(report))
}

/// Clears the unread badge a finished turn raised. The client that read the
/// node calls this; everyone else learns through the broadcast.
///
/// A focused client fires this on its own whenever a turn ends while the node is
/// on screen, so most calls arrive for a node that is already read. Those are
/// answered normally but not broadcast: re-announcing an unchanged row would put
/// one pointless frame on every workspace socket per finished turn.
pub async fn mark_agent_status_read(
    State(state): State<AppState>,
    AxumPath(node_id): AxumPath<String>,
) -> AppResult<Json<AgentStatus>> {
    // Once the agent domain has moved, this row is the Host's record and the
    // badge is cleared there (business migration §2.7). Reading a status keeps
    // answering from here, which is what makes the switch reversible: this
    // table stays the rollback baseline until migration 0012 retires it.
    crate::ownership::require_local_write(&state.pool, crate::ownership::OwnershipDomain::Agent)
        .await?;
    let receipt = db::mark_agent_status_read(&state.pool, &node_id)
        .await?
        .ok_or_else(|| AppError::NotFound("This node has never reported".into()))?;
    if receipt.cleared {
        state.events.publish(
            &receipt.status.workspace_id,
            WorkspaceEvent::AgentStatus {
                status: receipt.status.clone(),
            },
        );
    }
    Ok(Json(receipt.status))
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TranscriptQuery {
    max_bytes: Option<u64>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TranscriptExcerpt {
    pub node_id: String,
    pub text: String,
    /// The excerpt stopped at the rendered ceiling. A cut-off conversation read
    /// as a whole one is a wrong answer, not a short one.
    pub truncated: bool,
}

/// `GET /api/agent-status/{nodeId}/transcript` — the node's own conversation,
/// rendered one prose line per message.
///
/// This is a read, so it answers whichever side owns the agent domain's
/// records: the transcript is a file on this machine either way, and the
/// Worker channel's `ReadTranscript` (§2.7) reads the very same one.
///
/// A provider that keeps nothing readable is **501, not an empty body**.
/// OpenCode's history lives in a store only its own CLI exports, Pi and Oh My
/// Pi report a live context window rather than a file, and Copilot's
/// `events.jsonl` is an event log this renderer cannot turn into messages. An
/// empty excerpt would be indistinguishable from a session that has said
/// nothing yet, and the panel would draw the blank as the truth.
pub async fn read_agent_transcript(
    State(state): State<AppState>,
    AxumPath(node_id): AxumPath<String>,
    Query(query): Query<TranscriptQuery>,
) -> AppResult<Json<TranscriptExcerpt>> {
    let status = db::get_agent_status(&state.pool, &node_id)
        .await?
        .ok_or_else(|| AppError::NotFound("This node has never reported".into()))?;
    let provider = state.settings.base_agent(&status.agent_id);
    let located = crate::collab::transcript::locate(
        &provider,
        status.transcript_path.as_deref(),
        status.session_id.as_deref(),
    )
    .ok_or_else(|| {
        AppError::Unsupported(format!(
            "{provider} keeps no transcript this machine can read"
        ))
    })?;
    let budget = query
        .max_bytes
        .unwrap_or(crate::collab::transcript::MAX_TAIL_BYTES)
        .clamp(1, crate::collab::transcript::MAX_TAIL_BYTES);
    let text = crate::collab::transcript::read_tail(&located.path, budget)
        .map_err(|_| AppError::NotFound("The transcript could not be read".into()))?;
    let lines = crate::collab::transcript::render(&text);
    if lines.is_empty() {
        return Err(AppError::Unsupported(format!(
            "The file {provider} reports is not a conversation this reader renders"
        )));
    }
    let rendered = lines.join("\n");
    let ceiling = crate::collab::transcript::MAX_RENDERED_BYTES;
    let truncated = rendered.len() > ceiling;
    let text = if truncated {
        let mut end = ceiling;
        while end > 0 && !rendered.is_char_boundary(end) {
            end -= 1;
        }
        rendered[..end].to_string()
    } else {
        rendered
    };
    Ok(Json(TranscriptExcerpt {
        node_id,
        text,
        truncated,
    }))
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SuggestedTitle {
    pub title: String,
    /// Where the sentence came from, so the UI can be honest when the answer is
    /// only the agent's name: `transcript` / `terminal` / `agent`.
    pub source: &'static str,
}

/// `POST /api/agent-status/{nodeId}/suggest-title` — the header's ✦ button
/// (plan §17). Three sources, best first:
///
///   1. the transcript's first user message — what the session is *about*;
///   2. the last command in the pane, for a terminal that never reported one;
///   3. the agent's label, which is always available and never wrong.
///
/// No model is called: this is a rename button, and a local read answers it in
/// milliseconds without spending a token.
pub async fn suggest_agent_title(
    State(state): State<AppState>,
    AxumPath(node_id): AxumPath<String>,
) -> AppResult<Json<SuggestedTitle>> {
    let status = db::get_agent_status(&state.pool, &node_id)
        .await?
        .ok_or_else(|| AppError::NotFound("This node has never reported".into()))?;

    if let Some(path) = status.transcript_path.as_deref()
        && let Some(title) = index::transcript_title(&status.agent_id, Path::new(path))
    {
        return Ok(Json(SuggestedTitle {
            title,
            source: "transcript",
        }));
    }

    // The node's terminal keeps its logical key across recycles, so the lookup
    // is by node id rather than by the session id the status row happens to
    // remember.
    if let Ok(session) = db::get_terminal_session_by_key(&state.pool, &node_id).await
        && let Ok(capture) = state.terminals.capture(&session.id, 40, false).await
        && let Some(title) = index::command_from_capture(&capture.data)
    {
        return Ok(Json(SuggestedTitle {
            title,
            source: "terminal",
        }));
    }

    Ok(Json(SuggestedTitle {
        title: agent::definition(&status.agent_id)
            .map(|agent| agent.label.to_owned())
            .unwrap_or_else(|| status.agent_id.clone()),
        source: "agent",
    }))
}
